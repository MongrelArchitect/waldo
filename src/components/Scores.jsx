import React, { useEffect, useState } from 'react';
import { collection, getDocs } from 'firebase/firestore';
import db from '../helpers/firebase';
import formatTime from '../helpers/timer';

export default function Scores({ timer }) {
  const [scores, setScores] = useState([]);

  useEffect(() => {
    clearInterval(timer);
    const setFromDatabase = async () => {
      const newScores = [];
      const querySnapshot = await getDocs(collection(db, 'scores'));
      querySnapshot.forEach((doc) => {
        newScores.push({ name: doc.id, time: doc.data().time });
      });
      setScores(newScores.sort((a, b) => a.time - b.time));
    };
    setFromDatabase();
  }, []);

  return (
    <div className="scores">
      <h2>Scoreboard</h2>
      <table>
        <thead>
          <tr>
            <th>Rank</th>
            <th>Name</th>
            <th>Time</th>
          </tr>
        </thead>
        <tbody aria-label="table body">
          {scores.map((score) => (
            <tr key={score.name}>
              <td>{scores.indexOf(score) + 1}</td>
              <td>{score.name}</td>
              <td>{formatTime(score.time)}</td>
            </tr>
          ))}
        </tbody>
      </table>
    </div>
  );
}
