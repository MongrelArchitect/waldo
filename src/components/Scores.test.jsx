import React from 'react';
import { render, screen } from '@testing-library/react';
import Scores from './Scores';

describe('Scores component', () => {
  it('exists', () => {
    render(<Scores />);
    expect(screen.getByRole('table')).toBeTruthy();
  });

  it('displays correct number of headings', () => {
    render(<Scores />);
    expect(screen.getAllByRole('columnheader').length).toBe(3);
  });

  it('matches snapshot', () => {
    const { container } = render(<Scores />);
    expect(container).toMatchSnapshot();
  });
});
