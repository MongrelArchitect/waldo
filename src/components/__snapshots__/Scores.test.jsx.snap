// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Scores component matches snapshot 1`] = `
<div>
  <div>
    <table>
      <thead>
        <tr>
          <th>
            Rank
          </th>
          <th>
            Name
          </th>
          <th>
            Time
          </th>
        </tr>
      </thead>
      <tbody
        aria-label="table body"
      />
    </table>
  </div>
</div>
`;
